### PR TITLE
feat: ephemeral tables

### LINKED ISSUE #342: Proposal: Actions
_This is mostly in an idea phase rather than a well-formed proposal. Hoping to start a discussion around use cases and potential solutions._

There are many cases in which you might want to know when certain actions are being taken by players, that may or may not result in component updates, e.g. a player attacks another player and misses.

We have a system call stream, but it's just a different view on component updates (grouped by system calls), so it depends on a component updates being emitted for the system call to propagate to clients.

This also means that you have to infer what happened based on component updates, essentially duplicating logic on the client that already exists in the contracts (source of truth). "Attack system was called, and player health lowered, so they must have landed a hit and taken damage." We could expand the system call stream to include _all_ system calls, but it still requires inferring behavior in the client, and streaming this data from the RPC node isn't as efficient as filtering on events.

For cases where you might want to "listen" to things like a "missed attack", you can create/update an arbitrary component to get it into the system call stream, and then check that no damage was done (or however you model your combat system), but this feels more like a workaround and adds a lot more gas.

I'd like to propose we have some generic Solidity event type that you can emit from systems (and/or is emitted automatically on `execute` calls) that can represent named actions. Then expose an action stream so that UIs can trigger animations, effects, etc. We can create some patterns that help guide folks away from one-off modeling and preferring ECS where possible, but I think the cases above just can't be modeled well in our current setup with system call stream, etc.

### PR DESCRIPTION
**Pull request by dk1a**
Closes #342 
Docs in #758

### CODE REVIEW

**Comment by Kooshaba** [comment]
this is the minimal version of ephemeral tables needed so we can start using it in examples.

question: is it a problem if we keep the new events as part of the base world for now? or should i already move it out into a module?

client integration does not cache or replay ephemeral events on load (do not want to replay all animations associated with them).

File: examples/v2-minimal/packages/contracts/mud.config.mts
Comment (holic) [inline examples/v2-minimal/packages/contracts/mud.config.mts:21]: I'm a little confused by this example. A `bool` value type but the client renders as a counter. Is there another example or name we could use that might make this more clear?

File: examples/v2-minimal/packages/contracts/mud.config.mts
Comment (holic) [inline examples/v2-minimal/packages/contracts/mud.config.mts:23]: Still not entirely sold on "ephemeral" as the name for this pattern but happy to go with this for now (haven't come up with anything better)

File: examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol
Comment (holic) [inline examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol:66]: wondering if `registerSchema` and `setMetadata` should also be ephemeral (only events) or not

feels weird to write these table definitions even though the underlying tables will never be written to

File: examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts
Comment (holic) [inline examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts:60]: @alvrs @dk1a I think the answer is no, but does our module system support opt-in features like this but without the need to go through the hooks path? Just worried about extending the base world with ephemeral events if we're running low on contract size and have other array methods planned.

File: examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts
Comment (holic) [inline examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts:504]: we should probably align the names of these
- `StoreEphemeralSetRecord` event
- `setEphemeralRecord` function

File: packages/config/src/store/parseStoreConfig.ts
Comment (holic) [inline packages/config/src/store/parseStoreConfig.ts:74]: Hmm, rather than reducing the schema to just one `setEphemeral` field, I wonder if we can support any/all fields, with their underlying implementation calling out to `setEphemeralRecord`?

A big reason I was initially excited by the ephemeral table design is that it could look/act like a regular table (thus multiple fields), except only emit events for records rather than all the write methods.

File: examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts
Comment (holic) [reply examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts:504]: in terms of API, I am wondering if it would be more clear to use the term "emit" instead of "set" and just call it an ephemeral record, e.g.
- `StoreEphemeralRecord` (since there is no notion of set/delete/field updates, etc)
- `emitEphemeralRecord` (rather than set)

File: packages/network/src/types.ts
Comment (holic) [inline packages/network/src/types.ts:133]: do we need a separate event for this? why can't we lean on `NetworkComponentUpdate` as it is? maybe so we can discern whether or not this record should be cached downstream?

**Review by holic — COMMENTED** [review]
Mostly left some high level thoughts/comments on API design.

If this is blocking hackathon things, I'm fine with landing this as is if we set the expectation that this API is likely to change before v2 stable release.

I'll defer to @alvrs for making a call on merging this as is vs. continuing to iterate.

File: packages/network/src/types.ts
Comment (Kooshaba) [reply packages/network/src/types.ts:133]: yes its purely for caching. i agree it's a bit dumb.

File: examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts
Comment (Kooshaba) [reply examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts:504]: i agree with emit

File: examples/v2-minimal/packages/contracts/mud.config.mts
Comment (Kooshaba) [reply examples/v2-minimal/packages/contracts/mud.config.mts:23]: i've been thinking about it and also haven't come up with anything

File: examples/v2-minimal/packages/contracts/mud.config.mts
Comment (Kooshaba) [reply examples/v2-minimal/packages/contracts/mud.config.mts:21]: the increment system example is super contrived. the idea was that it represents that something happened (in this case with no information because literally nothing can happen)

File: packages/network/src/types.ts
Comment (holic) [reply packages/network/src/types.ts:133]: I added `partialValue` and `initialValue` to `NetworkComponentUpdate` for the purposes of v2. Another route is to extend `NetworkComponentUpdate` with something like `cache?: boolean` where it defaults to true and you can explicitly set to `false` for ephemeral events.

File: examples/v2-minimal/packages/contracts/mud.config.mts
Comment (ludns) [reply examples/v2-minimal/packages/contracts/mud.config.mts:23]: ask chatgpt

File: packages/config/src/store/parseStoreConfig.ts
Comment (dk1a) [reply packages/config/src/store/parseStoreConfig.ts:74]: The problem is probably that my comment was misleading, ephemeral option does allow multiple fields, it just removed ass `set` functions, and adds a new `setEphemeral`, which functions like `setRecord` but with a different event and no storage writes. `setEphemeral` supports the same variety/number of fields as `setRecord`

File: examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts
Comment (dk1a) [reply examples/v2-minimal/packages/contracts/types/ethers-contracts/IWorld.ts:60]: yeah if you mean the as-yet-unimplemented plugin hooks, modules don't need them for this. `core` module is a good example, as all it does is add more functions to `World` because of contract size issues

File: examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol
Comment (dk1a) [reply examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol:66]: that might be bad for sync, snap sync in particular. Also complicates tablegen a bit

Comment (alvrs) [reply examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol:66]: would tent to keeping this, since it might make it easier to sync the ephemeral events to off-chain indexers if this is ever wanted

File: examples/v2-minimal/packages/contracts/mud.config.mts
Comment (alvrs) [reply examples/v2-minimal/packages/contracts/mud.config.mts:23]: I would have proposed "transient" table, but I feel like we'll need this name for tables that use Solidity's upcoming transient storage

File: packages/config/src/store/parseStoreConfig.ts
Comment (alvrs) [reply packages/config/src/store/parseStoreConfig.ts:74]: If we go with `emit` (https://github.com/latticexyz/mud/pull/665#discussion_r1183035199) we could generate all `set*` with `emit*` and change the comment to something like "Generate a table that only emits `SetEphemeralRecord` events but doesn't write to storage."

File: packages/network/src/v2/ecsEventFromLog.ts
Comment (alvrs) [inline packages/network/src/v2/ecsEventFromLog.ts:40]: why do we need to cast here?

File: packages/network/src/types.ts
Comment (alvrs) [reply packages/network/src/types.ts:133]: would it work if we add `ephemeral: boolean` to `NetworkComponentUpdate` instead of a new event? Seems like that would simplify a lot of the logic/types below

(would tend to `ephemeral` instead of `cache` bc it would be used both for caching inside the sync logic, but also for downstream consumers like in `setupMUDV2Network`/ `applyNetworkUpdates`)

File: packages/network/src/workers/syncUtils.ts
Comment (alvrs) [inline packages/network/src/workers/syncUtils.ts:358]: why `lodash` instead of the built in `Array.filter` (which also supports type guards)?

```suggestion
    events.filter(isNetworkComponentUpdateEvent)
```

File: packages/store/src/StoreCore.sol
Comment (alvrs) [inline packages/store/src/StoreCore.sol:374]: Do we need to add `setEphemeralField` too?

File: packages/world/src/World.sol
Comment (alvrs) [inline packages/world/src/World.sol:170]: I think we should move this to a module to avoid increasing the World contract's bytecode size - can even be the existing `CoreModule` (note that nothing changes for consumers since a root function selector is registered)

File: packages/world/src/World.sol
Comment (alvrs) [inline packages/world/src/World.sol:275]: same here, should move this to `CoreModule`

File: examples/v2-minimal/packages/contracts/mud.config.mts
Comment (alvrs) [reply examples/v2-minimal/packages/contracts/mud.config.mts:21]: agree an example where some explicit data is "set" to the ephemeral table and then displayed in the client would be a bit clearer. Maybe something like an input field that forwards a string to a system, which sets the ephemeral table?

File: examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol
Comment (alvrs) [reply examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol:66]: I guess `deleteRecord` is not really needed in this lib though (and a bit misleading)

**Comment by alvrs** [comment]
> If this is blocking hackathon things, I'm fine with landing this as is if we set the expectation that this API is likely to change before v2 stable release.
>
>I'll defer to @alvrs for making a call on merging this as is vs. continuing to iterate.

Would prefer to resolve the open points in this PR before merging, but I think we're almost there! The open comments are mostly small things. Maybe except for the naming discussion, which we can iterate on later if we can't come up with a good alternative to "ephemeral" now.

File: examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol
Comment (dk1a) [reply examples/v2-minimal/packages/contracts/src/codegen/tables/EventTable.sol:66]: yeah, I missed deleteRecord because it has a separate rendering condition from the other record/field methods

File: packages/store/src/StoreCore.sol
Comment (dk1a) [reply packages/store/src/StoreCore.sol:374]: I'd favor only record, by analogy to events it'd be like emitting an event with partial arguments, which probably means you just want more events, or an argument in the event itself to specify which arguments to look at

File: packages/network/src/types.ts
Comment (Kooshaba) [reply packages/network/src/types.ts:133]: ok wil add ephemeral 👍🏼

File: packages/store/src/IStore.sol
Comment (holic) [inline packages/store/src/IStore.sol:61]: Similar to above, if we think we'll have other operations (e.g. delete), then I think this is fine as is.

If we're only planning to have set (i.e. `emit StoreEphemeralRecord`), then I think it'd be clearer to have this as `emitEphemeralRecord` to indicate we're not storing anything.

File: packages/store/src/IStore.sol
Comment (dk1a) [reply packages/store/src/IStore.sol:58]: I like `StoreEphemeralRecord`, and don't really expect a need for "delete"

File: examples/minimal/packages/contracts/src/systems/ChatSystem.sol
Comment (alvrs) [inline examples/minimal/packages/contracts/src/systems/ChatSystem.sol:13]: The `SingletonKey` pattern was confusing to folks at the last hackathon (esp. since tables also support `singleton: true` for "real singleton tables"). Should we just use the `msg.sender` as key? Then the frontend could even display the sender of the message.

File: packages/network/src/v2/common.ts
Comment (alvrs) [inline packages/network/src/v2/common.ts:8]: Reminder to also change this if we change the event name to `StoreEphemeralRecord`

File: packages/network/src/v2/ecsEventFromLog.ts
Comment (alvrs) [inline packages/network/src/v2/ecsEventFromLog.ts:52]: just another reminder
```suggestion
  if (name === "StoreEphemeralRecord") {
    const value = await decodeStoreSetRecord(contract, tableId, args.key, args.data);
    console.log("StoreEphemeralRecord:", { table: tableId.toString(), component, entity, value });
    return {
      ...ecsEvent,
      ephemeral: true,
      value,
    };
  }

```

File: packages/store/src/StoreCore.sol
Comment (alvrs) [inline packages/store/src/StoreCore.sol:374]: let's call this `emitEphemeralRecord`?

**Review by alvrs — COMMENTED** [review]
Only tiny nits left - mainly let's rename `setEphemeralRecord` to `emitEphemeralRecord` and `StoreEphemeralSetRecord` to `StoreEphemeralRecord`

File: examples/minimal/packages/contracts/src/systems/ChatSystem.sol
Comment (ludns) [reply examples/minimal/packages/contracts/src/systems/ChatSystem.sol:13]: I think we should stop exposing the SingletonKey to user and surface APIs contract and client side to interact with Singleton (we already do have the API tablegen side, we need it RECS / mud-react side too)

File: examples/minimal/packages/contracts/src/systems/ChatSystem.sol
Comment (alvrs) [inline examples/minimal/packages/contracts/src/systems/ChatSystem.sol:6]:
```suggestion
```
No need for this anymore

File: packages/store/ts/library/render-solidity/renderTable.ts
Comment (alvrs) [inline packages/store/ts/library/render-solidity/renderTable.ts:35]: For consistency we should add a `withDeleteMethods` field to the config (like `withFieldMethods`, `withRecordMethods`, `withEphemeralMethods`). Not strictly blocking for this PR, we can also do that in a follow up.

File: packages/world/src/modules/core/implementations/EphemeralRecordSystem.sol
Comment (alvrs) [inline packages/world/src/modules/core/implementations/EphemeralRecordSystem.sol:14]:
```suggestion
```
it seems like we're not doing anything in here that would require this

File: packages/world/src/modules/core/implementations/EphemeralRecordSystem.sol
Comment (alvrs) [inline packages/world/src/modules/core/implementations/EphemeralRecordSystem.sol:15]:
```suggestion

contract EphemeralRecordSystem is System {
```
it seems like we're not doing anything in here that would require this

**Review by alvrs — APPROVED** [review]
just one nit left (https://github.com/latticexyz/mud/pull/665#discussion_r1187909532) but not blocking. Good to go imo!

**Review by alvrs — APPROVED** [review]
just one nit left (https://github.com/latticexyz/mud/pull/665#discussion_r1187909532) but not blocking. Good to go imo!

**Pull request merged**
